numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -985,6 +985,72 @@ PyArray_CanCastArrayTo(PyArrayObject *arr, PyArray_Descr *to,
     return PyArray_CanCastTypeTo(from, to, casting);
 }
 
+
+NPY_NO_EXPORT const char *
+npy_casting_to_string(NPY_CASTING casting)
+{
+    switch (casting) {
+        case NPY_NO_CASTING:
+            return "'no'";
+        case NPY_EQUIV_CASTING:
+            return "'equiv'";
+        case NPY_SAFE_CASTING:
+            return "'safe'";
+        case NPY_SAME_KIND_CASTING:
+            return "'same_kind'";
+        case NPY_UNSAFE_CASTING:
+            return "'unsafe'";
+        default:
+            return "<unknown>";
+    }
+}
+
+
+/**
+ * Helper function to set a useful error when casting is not possible.
+ *
+ * @param src_dtype
+ * @param dst_dtype
+ * @param casting
+ * @param scalar Whether this was a "scalar" cast (includes 0-D array with
+ *               PyArray_CanCastArrayTo result).
+ */
+NPY_NO_EXPORT void
+npy_set_invalid_cast_error(
+        PyArray_Descr *src_dtype, PyArray_Descr *dst_dtype,
+        NPY_CASTING casting, npy_bool scalar)
+{
+    PyObject *errmsg;
+    PyObject *src_dtype_repr = NULL;
+    PyObject *dst_dtype_repr = NULL;
+
+    if (!scalar) {
+        errmsg = PyUString_FromString("Cannot cast array data from ");
+    }
+    else {
+        errmsg = PyUString_FromString("Cannot cast scalar from ");
```

Comment:
Python 3 has a much shorter way to write all this, with %R:
```suggestion
        PyErr_Format(PyExc_TypeError, "Cannot cast scalar from %R to %R according to the rule %s");
```

Reply:
Good point, didn't really try to fix up that code, had just moved it.